lib/jsdom/living/fetch/Headers-impl.js

```diff
@@ -0,0 +1,231 @@
+  switch (name) {
+    case "accept":
+      if (value.match(CORS_UNSAFE_BYTE)) {
+        return false;
+      }
+      break;
+    case "accept-language":
+    case "content-language":
+      if (value.match(/[^\x30-\x39\x41-\x5A\x61-\x7A *,\-.;=]/)) {
+        return false;
+      }
+      break;
+    case "content-type": {
+      if (value.match(CORS_UNSAFE_BYTE)) {
+        return false;
+      }
+      const mimeType = MIMEType.parse(value);
+      if (mimeType === null) {
+        return false;
+      }
+      if (
+        ![
+          "application/x-www-form-urlencoded",
+          "multipart/form-data",
+          "text/plain"
+        ].includes(mimeType.essence)
+      ) {
+        return false;
+      }
+      break;
+    }
+    default:
+      return false;
+  }
+  if (Buffer.from(value).length > 128) {
+    return false;
+  }
+  return true;
+}
+
+class HeadersImpl {
+  constructor(args) {
+    this.guard = "none";
+    this.headersList = new Map();
+
+    if (args[0]) {
+      this._fill(args[0]);
+    }
+  }
+
+  _fill(init) {
+    if (Array.isArray(init)) {
+      for (const header of init) {
+        if (header.length !== 2) {
+          throw new TypeError("init is invalid");
+        }
+        this.append(header[0], header[1]);
+      }
+    } else {
+      for (const key of Object.keys(init)) {
+        this.append(key, init[key]);
+      }
+    }
+  }
+
+  has(name) {
+    assertName(name);
+    return this.headersList.has(name.toLowerCase());
+  }
+
+  get(name) {
+    assertName(name);
+    const values = this.headersList.get(name.toLowerCase());
+    if (!values) {
+      return null;
+    }
+    return values.map(h => h.value).join(", ");
+  }
+
+  _removePrivilegedNoCORSHeaders() {
+    this.headersList.delete("range");
+  }
+
+  _append(name, value) {
+    const existing = this.headersList.get(name.toLowerCase());
+    if (existing) {
+      name = existing[0].name;
+      existing.push({ name, value });
+    } else {
+      this.headersList.set(name.toLowerCase(), [{ name, value }]);
+    }
+  }
+
+  append(name, value) {
+    value = value.trim();
+    assertName(name);
+    assertValue(value);
+
+    switch (this.guard) {
+      case "immutable":
+        throw new TypeError("Headers is immutable");
+      case "request":
+        if (isForbiddenHeader(name)) {
+          return;
+        }
+        break;
+      case "request-no-cors": {
+        let temporaryValue = this.get(name);
+        if (temporaryValue === null) {
+          temporaryValue = value;
+        } else {
+          temporaryValue += `, ${value}`;
+        }
+        if (!isCORSWhitelisted(name, value)) {
+          return;
+        }
+        break;
+      }
+      case "response":
+        if (FORBIDDEN_RESPONSE.has(name.toLowerCase())) {
+          return;
+        }
+        break;
+    }
+
+    this._append(name, value);
+    this._removePrivilegedNoCORSHeaders();
+  }
+
+  set(name, value) {
+    value = value.trim();
```

Comment:
trim() will remove Unicode whitespace... I guess it's OK though since we're behind a ByteString guard, interesting.

Reply:
I see use of toLowerCase() as well. Perhaps a comment at the top of the file explaining that it's OK to use normally-Unicode string operations because the bindings layer guarantees that the strings are ByteStrings.